python/ray/data/dataset.py

```diff
@@ -845,15 +846,115 @@ def groupby(self, key: "GroupKeyT") -> "GroupedDataset[T]":
         from ray.data.grouped_dataset import GroupedDataset
         return GroupedDataset(self, key)
 
-    def aggregate(self, *aggs: AggregateFn) -> U:
-        """Aggregate the entire dataset as one group.
+    def agg(self, aggs: Dict[AggregateOnT, List["Aggregation"]]):
+        """Aggregate using one or more operations over the specified data
+        subsets.
+
+        Common aggregations can be given as strings (``"sum"``, ``"min"``,
+        ``"max"``, ``"mean"``, ``"std"``) while custom/customized aggregations
+        can be given as instances of :class:`~.AggregateFn`. If providing a
+        dictionary of data subsets --> aggregations, the data subset keys can
+        be column name strings for table (Arrow) datasets, and can be
+        row --> data subset callables for both table (Arrow) datasets and
+        simple datasets. See examples below.
+
+        This is a blocking operation.
+
+        Examples:
+            >>> ray.data.range(100).agg({
+            ...     lambda x: x % 3: ["min", "max"],
+            ...     lambda x: x % 2: ["mean", "std"],
```

Comment:
Could we provide a simpler example too, like "lambda x: x" in the common case? I think it would be nice to have "None" here as an alias for "lambda x: x".

Reply:
I had proposed those semantics for a `None` key earlier, and we settled on removing that logic. Should I add it back?

Reply:
I'm ok with it as an alias for "identity lambda". I think this is helpful in case you want to implement a custom aggregate over the entire ArrowRow as well.

I'd push back on "expanding to all fields" though, that's pretty strange unless you're coming from pandas.

Reply:
So would `None` only be a valid key for simple datasets, and we'd throw an error if it's given for an Arrow dataset?

Reply:
It can return ArrowRow for Arrow dataset (you might want to aggregate over some complex combination of columns, which would otherwise not be possible without this escape-hatch).